cloud/aws/lambda-handler.py

```diff
@@ -2,27 +2,67 @@
 import subprocess
 import logging
 import base64
+from threading import Thread
 
-logging.basicConfig(level=logging.INFO)
+logging.getLogger().setLevel(logging.INFO)
+
+
+class ReturningThread(Thread):
+    """A wrapper around the Thread class to actually return the threaded function
+    return value when calling join()"""
+
+    def __init__(
+        self, group=None, target=None, name=None, args=(), kwargs={}, Verbose=None
+    ):
+        Thread.__init__(self, group, target, name, args, kwargs)
+        self._return = None
+
+    def run(self):
+        if self._target is not None:
+            self._return = self._target(*self._args, **self._kwargs)
+
+    def join(self, *args):
+        Thread.join(self, *args)
+        return self._return
+
+
+def buff_and_print(stream, stream_name):
+    """Buffer and log every line of the given stream"""
+    buff = []
+    for l in iter(lambda: stream.readline(), b""):
+        line = l.decode("utf-8")
+        logging.info("%s: %s", stream_name, line.rstrip())
+        buff.append(line)
+    return "".join(buff)
 
 
 def handler(event, context):
     """An AWS Lambda handler that runs the provided command with bash and returns the standard output"""
     try:
-        logging.info("event: ", event)
+        # input parameters
+        logging.debug("event: %s", event)
         src_cmd = base64.b64decode(event["cmd"]).decode("utf-8")
         host = event["host"]
+        logging.info("src_cmd: %s", src_cmd)
+        logging.info("host: %s", host)
 
         # execute the command as bash and return the std outputs
         parsed_cmd = ["/bin/bash", "-c", src_cmd]
         os.environ["VAST_ENDPOINT"] = host
         process = subprocess.Popen(
             parsed_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE
         )
-        stdout, stderr = process.communicate()
-        res = stdout if stdout != b"" else stderr
+        # we need to spin up a thread to avoid deadlock when reading through output pipes
+        stderr_thread = ReturningThread(
+            target=buff_and_print, args=(process.stderr, "stderr")
+        )
+        stderr_thread.start()
+        stdout = buff_and_print(process.stdout, "stdout")
+        stderr = stderr_thread.join()
 
-        return {"result": res.decode("utf-8"), "parsed_cmd": parsed_cmd}
+        # multiplex stdout and stderr into the result field
+        res = stdout if stdout != "" else stderr
```

Comment:
would it be desirable to have both stdout _and_ stderr in the output?

Reply:
Yes, but I don't really know how to format it. One logical path would be to return both stdout and stderr in the result object. But then you need a bit more formatting on the client side. Do you have a suggestion?

Reply:
I'm not even sure if there's a sensible combination of stderr and stdout. Let's handle that "later" when we see the need.